app/naive_bayes.py

```diff
@@ -0,0 +1,286 @@
+                    contact_type, trusted_status, privilege, architecture = dict_agent_paw_to_info[agent_paw]
+                    dict_link_features["Agent_Protocol"] = contact_type
+                    dict_link_features["Trusted_Status"] = trusted_status
+                    dict_link_features["Agent_Privilege"] = privilege
+                    dict_link_features["Host_Architecture"] = architecture
+                else:
+                    dict_link_features["Agent_Protocol"] = None
+                    dict_link_features["Trusted_Status"] = None
+                    dict_link_features["Agent_Privilege"] = None
+                    dict_link_features["Host_Architecture"] = None
+
+                dict_link_features["Link_Facts"] = self._get_useful_facts(cur_link)
+
+                link_row = [
+                    dict_link_features["Status"],
+                    dict_link_features["Ability_ID"], 
+                    dict_link_features["Link_Facts"], 
+                    dict_link_features["Planner"],
+                    dict_link_features["Obfuscator"],
+                    dict_link_features["Adversary_ID"],
+                    dict_link_features["Adversary_Name"],
+                    dict_link_features["Command"],
+                    dict_link_features["Number_Facts"],
+                    dict_link_features["Visibility_Score"],
+                    dict_link_features["Executor_Platform"],
+                    dict_link_features["Executor_Name"],
+                    dict_link_features["Agent_Protocol"],
+                    dict_link_features["Trusted_Status"],
+                    dict_link_features["Agent_Privilege"],
+                    dict_link_features["Host_Architecture"]
+                ]
+                matrix_link_data.append(link_row)
+        return matrix_link_data
+
+    def _get_useful_facts(self, cur_link : Link) -> Dict[str, str]:
+        """
+        Selects generalizable facts from link object for data querying
+        """
+        useful_facts = {}
+        if(len(cur_link.used) > 0):
+            for used_fact in cur_link.used:
+                # exclude fact types that are unique to a specific host
+                if str(used_fact.trait).startswith(("host.", "remote.", "file.last.", "domain.user.")) == False:
+                    useful_facts[str(used_fact.trait)] = str(used_fact.value)
+        return useful_facts
+
+    def _query_link_matrix(self, matrix_link_data : List[List], query_features : Dict) -> List[List]:
+        """
+        Query link matrix for links containing specific features and values and 
+        return new matrix composed of column names row and the link rows containing right feat and val
+
+        :param matrix_link_data: link matrix that will be queried
+        :param query_features: dict of {feature_names: values} that matrix_link_data will be queried by
+        """
+        col_name_to_col_index = {}
+        list_col_names = matrix_link_data[0]
```

Comment:
Is it guaranteed that `matrix_link_data` has a length greater than zero?

Reply:
Oh nevermind, I see. The first element of `matrix_link_data` is the list of feature names. If you instead define that list of feature names as a constant at the top of the file, you could use that constant instead of having to grab it off the top of the list every time. I would be wary of having your list of `matrix_link_data` include both the feature names as its first element and actual feature data as the rest of its elements. It's hard to convey that in a way that's easily understandable to someone else reading the code. It would be preferable to have `matrix_link_data` only include data and not include the feature name list at all. If the feature names are defined at the module scope as a constant, you wouldn't need to include it in this list anyway.